File: backend/app/schemas/compliance.py
"""
Phase 4 (+ Rule Engine v2): Compliance schemas for the Legal Metrology Rules Engine.
"""

from typing import List, Optional, Any
from pydantic import BaseModel
from enum import Enum


class ComplianceStatus(str, Enum):
    PASS = "PASS"
    FAIL = "FAIL"
    REVIEW = "REVIEW"
    NOT_APPLICABLE = "NOT_APPLICABLE"
    NOT_ANALYSED = "NOT_ANALYSED"


class RuleSeverity(str, Enum):
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class RuleCategory(str, Enum):
    """
    Report-facing grouping for findings. Deliberately a small, fixed set so the
    frontend can render a stable set of sections regardless of how many rules
    are registered in the engine.
    """
    MANDATORY_DECLARATIONS = "MANDATORY_DECLARATIONS"
    QUANTITY_UNITS = "QUANTITY_UNITS"
    PRICING_MRP = "PRICING_MRP"
    PRODUCT_SPECIFIC = "PRODUCT_SPECIFIC"
    PRESENTATION_LABEL = "PRESENTATION_LABEL"
    CONDITIONAL_ADDITIONAL = "CONDITIONAL_ADDITIONAL"


RULE_CATEGORY_LABELS = {
    RuleCategory.MANDATORY_DECLARATIONS: "Mandatory Declarations",
    RuleCategory.QUANTITY_UNITS: "Quantity & Units",
    RuleCategory.PRICING_MRP: "Pricing / MRP",
    RuleCategory.PRODUCT_SPECIFIC: "Product-Specific Requirements",
    RuleCategory.PRESENTATION_LABEL: "Presentation / Label Requirements",
    RuleCategory.CONDITIONAL_ADDITIONAL: "Additional / Conditional Checks",
}


class RuleResultSchema(BaseModel):
    rule_id: str
    rule_name: str
    status: ComplianceStatus
    severity: RuleSeverity
    category: RuleCategory = RuleCategory.MANDATORY_DECLARATIONS
    message: str
    field: Optional[str] = None
    expected: str
    actual: Optional[str] = None
    # evidence structure matching the ProductInfo schema
    evidence: Optional[Any] = None
    source_reference: str
    # Rule engine v2 additions — all optional so older persisted rows still validate.
    confidence: Optional[float] = None
    remediation: Optional[str] = None


class CategorySummarySchema(BaseModel):
    """Aggregated counts for one report category — powers the grouped findings UI."""
    category: RuleCategory
    label: str
    passed_count: int = 0
    failed_count: int = 0
    review_count: int = 0
    not_applicable_count: int = 0
    total: int = 0


class ComplianceReportSchema(BaseModel):
    inspection_id: str
    image_id: str
    overall_status: ComplianceStatus
    
    # Aggregated metrics
    total_rules_checked: int
    passed_count: int
    failed_count: int
    review_count: int
    not_applicable_count: int
    
    rule_results: List[RuleResultSchema] = []
    # Rule engine v2: precomputed grouping so the frontend doesn't need to
    # duplicate category logic.
    category_breakdown: List[CategorySummarySchema] = []


class ComplianceResponse(BaseModel):
    inspection_id: str
    overall_status: ComplianceStatus
    reports: List[ComplianceReportSchema] = []

File: frontend/src/api/compliance.ts
import client from './client';

export type ComplianceStatus = 'PASS' | 'FAIL' | 'REVIEW' | 'NOT_APPLICABLE';
export type RuleSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';

export interface RuleResult {
  rule_id: string;
  rule_name: string;
  status: ComplianceStatus;
  severity: RuleSeverity;
  message: string;
  field?: string;
  expected: string;
  actual?: string;
  evidence?: any;
  source_reference: string;
  category?: string;
  confidence?: number;
  remediation?: string;
  /** Which evidence image this specific finding's evidence came from. Set during merge. */
  image_id?: string;
}

export interface ComplianceReport {
  inspection_id: string;
  image_id: string;
  overall_status: ComplianceStatus;
  total_rules_checked: number;
  passed_count: number;
  failed_count: number;
  review_count: number;
  not_applicable_count: number;
  rule_results: RuleResult[];
  category_breakdown?: {
    category: string;
    label: string;
    passed_count: number;
    failed_count: number;
    review_count: number;
    not_applicable_count: number;
    total: number;
  }[];
}

export interface ComplianceResponse {
  inspection_id: string;
  overall_status: ComplianceStatus;
  reports: ComplianceReport[];
}

export const runComplianceAnalysis = async (inspectionId: string): Promise<ComplianceResponse> => {
  const response = await client.post<ComplianceResponse>(`/inspections/${inspectionId}/compliance`);
  return response.data;
};

export const getComplianceReports = async (inspectionId: string): Promise<ComplianceResponse> => {
  const response = await client.get<ComplianceResponse>(`/inspections/${inspectionId}/compliance`);
  return response.data;
};

/**
 * Consolidate the per-image compliance reports into one product-level verdict.
 *
 * The backend evaluates rules separately for each evidence image (one
 * ComplianceReport per photo), because a single photo often can't show every
 * mandatory declaration on the package. Rendering each per-image report
 * on its own produces contradictory-looking results for one physical
 * product — e.g. "Consumer contact: FAIL" from the front-label photo next
 * to "Consumer contact: PASS" from the back-label photo. This merges them
 * by rule_id so each requirement gets one verdict: PASS if satisfied on
 * ANY image, otherwise the least-bad outcome across the images checked.
 */
const STATUS_RANK: Record<ComplianceStatus, number> = {
  PASS: 0,
  NOT_APPLICABLE: 1,
  REVIEW: 2,
  FAIL: 3,
};

export function mergeComplianceReports(reports: ComplianceReport[]): ComplianceReport {
  if (reports.length === 1) return reports[0];

  const byRule = new Map<string, RuleResult>();
  for (const report of reports) {
    for (const result of report.rule_results) {
      const existing = byRule.get(result.rule_id);
      if (!existing || STATUS_RANK[result.status] < STATUS_RANK[existing.status]) {
        byRule.set(result.rule_id, { ...result, image_id: report.image_id });
      }
    }
  }

  const merged = Array.from(byRule.values());
  const counts: Record<ComplianceStatus, number> = { PASS: 0, FAIL: 0, REVIEW: 0, NOT_APPLICABLE: 0 };
  for (const r of merged) counts[r.status]++;

  type CategoryBucket = NonNullable<ComplianceReport['category_breakdown']>[number];
  const categoryMap = new Map<string, CategoryBucket>();
  for (const r of merged) {
    const cat = r.category || 'UNKNOWN';
    if (!categoryMap.has(cat)) {
      categoryMap.set(cat, { category: cat, label: cat, passed_count: 0, failed_count: 0, review_count: 0, not_applicable_count: 0, total: 0 });
    }
    const bucket = categoryMap.get(cat)!;
    bucket.total += 1;
    if (r.status === 'PASS') bucket.passed_count++;
    else if (r.status === 'FAIL') bucket.failed_count++;
    else if (r.status === 'REVIEW') bucket.review_count++;
    else bucket.not_applicable_count++;
  }
  // Carry over the human-readable category labels from whichever source report has them.
  for (const report of reports) {
    for (const cat of report.category_breakdown || []) {
      const bucket = categoryMap.get(cat.category);
      if (bucket) bucket.label = cat.label;
    }
  }

  // Mirrors the backend engine's own overall-status logic (see compliance/engine.py).
  const hasHighCriticalFail = merged.some((r) => r.status === 'FAIL' && (r.severity === 'CRITICAL' || r.severity === 'HIGH'));
  const overall_status: ComplianceStatus = hasHighCriticalFail
    ? 'FAIL'
    : counts.REVIEW > 0 || counts.FAIL > 0
    ? 'REVIEW'
    : 'PASS';

  return {
    inspection_id: reports[0].inspection_id,
    image_id: 'merged',
    overall_status,
    total_rules_checked: merged.length,
    passed_count: counts.PASS,
    failed_count: counts.FAIL,
    review_count: counts.REVIEW,
    not_applicable_count: counts.NOT_APPLICABLE,
    rule_results: merged,
    category_breakdown: Array.from(categoryMap.values()),
  };
}
